### PR TITLE
More comfortable entering of parent branches

### PR DESCRIPTION
**Pull request by kevgo**
@charlierudolph @allewun @ricmatsui 

This allows more efficient entry of the parent branch:
* shows a list of branches 
* allows to enter the name easier through copy-and-paste
* allows to enter the branch by number
* enter for main branch name

### CODE REVIEW

File: src/helpers/git_helpers/branch_configuration_helpers.sh [hunk 1]
```diff
@@ -104,9 +103,37 @@ function ensure_knows_parent_branches {
...
+        i=2
+        for branch in $branches; do
+          if [ "$branch" != "$current_branch" -a "$branch" != "$MAIN_BRANCH_NAME" ]; then
+            print_branch $i "$branch"
+            i=$(( i + 1 ))
+          fi
+        done
+
+        echo
+        echo -n "Please enter the parent branch name or number for $(echo_inline_cyan_bold "$current_branch") ($MAIN_BRANCH_NAME): "
```
Comment (allewun) [inline src/helpers/git_helpers/branch_configuration_helpers.sh]: How about:
> Please specify the parent branch of $current_branch by name or number: 

Also:
1. The ($MAIN_BRANCH_NAME) is a bit confusing just sitting there.
2. I've unfortunately not followed many of your recent changes closely. Is it clear to the user what a parent branch is? Maybe also having a brief one-liner explaining that (or a way for the user to read up on what this concept is)

File: src/helpers/git_helpers/branch_configuration_helpers.sh [hunk 2]
```diff
@@ -104,9 +103,37 @@ function ensure_knows_parent_branches {
...
-        if [ -z "$parent" ]; then
+        re='^[0-9]+$'
+        if [[ $parent =~ $re ]] ; then
+          # user entered a number here
+          parent=${branch_numbers[$parent]}
+        elif [ -z "$parent" ]; then
+          # user entered nothing
           parent=$MAIN_BRANCH_NAME
         fi
         if [ "$(has_branch "$parent")" == "false" ]; then
```
Comment (allewun) [inline src/helpers/git_helpers/branch_configuration_helpers.sh]: The error message for when a user types an invalid number is not very clear:

```
$ git sync

  1: master
  2: abc
  3: blah
  4: test

Please enter the parent branch name or number for unknown (master): 6

  Error
  branch '' doesn't exist
```

File: src/helpers/git_helpers/branch_configuration_helpers.sh [hunk 3]
```diff
@@ -104,9 +103,37 @@ function ensure_knows_parent_branches {
...
+
+        function print_branch {
+          local number=$1
+          local branch_name=$2
+
+          output_style_bold
+          printf "%3s: " "$number"
+          output_style_reset
+          echo "$branch_name"
+          branch_numbers[number]="$branch_name"
```
Comment (charlierudolph) [inline src/helpers/git_helpers/branch_configuration_helpers.sh]: I don't think this should happen inside this function. It can happen in the for loop and allows this method to be more focused.

Comment (charlierudolph) [inline src/helpers/git_helpers/branch_configuration_helpers.sh]: branch_numbers should also be declared a local variable then

Comment (kevgo) [inline src/helpers/git_helpers/branch_configuration_helpers.sh]: Good catch, both comments implemented

File: src/helpers/git_helpers/branch_configuration_helpers.sh (hunk 1 above)
Comment (kevgo) [inline src/helpers/git_helpers/branch_configuration_helpers.sh]: Good feedback! I have improved the wording according to your suggestion, and added the 10-second summary of branch nesting. Please apply the @allewun style and sanity checker over it again!  :)

File: src/helpers/git_helpers/branch_configuration_helpers.sh (hunk 2 above)
Comment (kevgo) [inline src/helpers/git_helpers/branch_configuration_helpers.sh]: Wow, thanks for testing this so thoroughly! I have added a better error message for this case now ("Error: Invalid branch number"). Good error messages are important!

**Comment by kevgo** [comment]
all comments addressed

File: src/helpers/git_helpers/branch_configuration_helpers.sh [hunk 4]
```diff
@@ -104,9 +103,53 @@ function ensure_knows_parent_branches {
...
         read parent
-        if [ -z "$parent" ]; then
+        re='^[0-9]+$'
+        if [[ $parent =~ $re ]] ; then
+          # user entered a number here
+          parent=${branch_numbers[$parent]}
+          if [ -z "$parent" ]; then
+            echo_error_header
+            echo_error "Invalid branch number"
+            exit_with_error newline
```
Comment (charlierudolph) [inline src/helpers/git_helpers/branch_configuration_helpers.sh]: Instead of exiting with error how about just asking again? This can be dealt in another PR.

File: features/git-sync/current_branch/feature_branch/nested_feature_branch/unknown_parent.feature [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    And the following commits exist in my repository
+      | BRANCH  | LOCATION         | MESSAGE        | FILE NAME    |
+      | main    | local and remote | main commit    | main_file    |
+      | feature | local and remote | feature commit | feature_file |
+    And I am on the "feature" branch
+    And I have an uncommitted file
+
+
+  Scenario: choosing the default branch name
+    When I run `git sync` and enter ""
```
Comment (charlierudolph) [inline features/git-sync/current_branch/feature_branch/nested_feature_branch/unknown_parent.feature]: After #563 this can be updated to `press ENTER`

File: src/helpers/git_helpers/branch_configuration_helpers.sh (hunk 4 above)
Comment (kevgo) [inline src/helpers/git_helpers/branch_configuration_helpers.sh]: Yes, good idea. #573 tracks this suggestion

File: features/git-sync/current_branch/feature_branch/nested_feature_branch/unknown_parent.feature (hunk 5 above)
Comment (kevgo) [inline features/git-sync/current_branch/feature_branch/nested_feature_branch/unknown_parent.feature]: Will do

**Comment by kevgo** [comment]
All comments addressed

File: features/git-sync/current_branch/feature_branch/nested_feature_branch/unknown_parent.feature (hunk 5 above)
Comment (kevgo) [inline features/git-sync/current_branch/feature_branch/nested_feature_branch/unknown_parent.feature]: Actually, #568, which implements #563, performs exactly this change, since it is a child branch of this one!

File: src/helpers/git_helpers/branch_configuration_helpers.sh (hunk 2 above)
Comment (kevgo) [inline src/helpers/git_helpers/branch_configuration_helpers.sh]: I have also added feature specs for entering a wrong number, or a non-existing branch name now.

File: src/helpers/git_helpers/branch_configuration_helpers.sh [hunk 6]
```diff
@@ -104,9 +103,53 @@ function ensure_knows_parent_branches {
...
+        print_branch 1 "$MAIN_BRANCH_NAME"
+        branch_numbers[1]=$MAIN_BRANCH_NAME
+        i=2
+        for branch in $branches; do
+          if [ "$branch" != "$current_branch" -a "$branch" != "$MAIN_BRANCH_NAME" ]; then
+            branch_numbers[i]=$branch
+            print_branch $i "$branch"
+            i=$(( i + 1 ))
+          fi
+        done
```
Comment (charlierudolph) [inline src/helpers/git_helpers/branch_configuration_helpers.sh]: I think we should try to break this function up as it has grown quite large. Mentioned it on another PR so it can be dealt with separately.

**Comment by charlierudolph** [comment]
:ship:

File: src/helpers/git_helpers/branch_configuration_helpers.sh [hunk 7]
```diff
@@ -104,9 +103,53 @@ function ensure_knows_parent_branches {
...
+          printf "%3s: " "$number"
+          output_style_reset
+          echo "$branch_name"
+        }
+
+        echo
+        echo "Feature branches can be branched directly off "
+        echo "$MAIN_BRANCH_NAME or from other feature branches."
+        echo
+        echo "The former allows to develop and ship features completely"
```
Comment (allewun) [inline src/helpers/git_helpers/branch_configuration_helpers.sh]: Just a few suggestions on the wording, take what you like from it:

> The former allows developing and shipping completely independent features.
> The latter facilitates working on features that depend on other unshipped features.

File: src/helpers/git_helpers/branch_configuration_helpers.sh (hunk 6 above)
Comment (kevgo) [inline src/helpers/git_helpers/branch_configuration_helpers.sh]: Agreed, its bothering me too, but I haven't found a good way to go about it yet. Will do this separately.

File: src/helpers/git_helpers/branch_configuration_helpers.sh (hunk 7 above)
Comment (kevgo) [inline src/helpers/git_helpers/branch_configuration_helpers.sh]: You get the price for saying things nice and concise! Thanks!
